Stats Step 5: Handle corner cases for old and new stats UI

anitaa1990 opened the pull request:
Fixes #1336. As defined in this master thread [here](https://github.com/woocommerce/woocommerce-android/issues/1199), this PR is Step 5 of the new Stats improvements and adds logic to handle corner cases that arise when supporting two different sets of UI for the Stats page.

### Implementation
- Added new custom UI class called `MyStoreStatsRevertedNoticeCard` to display a banner on top of the old stats if `wc-admin` is disabled on a site but the v4 stats UI is already displayed to the user.
- Added new custom UI class called `MyStoreStatsAvailabilityCard` to display a banner on top of the old stats if `wc-admin` is active on a site but the old stats UI is already displayed to the user.
- Added a new option in the Settings page called `Beta features`. This option will only be displayed if the `wc-admin` is available for the site. 
- Added a new fragment called `BetaFeaturesFragment` that allows a user to opt in/out of the new stats UI.
- Added new events to track the button clicks for the banners:
  - `SETTINGS_BETA_FEATURES_BUTTON_TAPPED` - when the user clicks on the Beta features section in the Settings page.
  - `SETTINGS_BETA_FEATURES_NEW_STATS_UI_TOGGLED` - when the user opts in or out of the new stats ui by toggling the settings page option.
  - `DASHBOARD_NEW_STATS_REVERTED_BANNER_DISMISS_TAPPED` - when dismiss button is clicked on the stats reverted banner.
  - `DASHBOARD_NEW_STATS_REVERTED_BANNER_LEARN_MORE_TAPPED` - when learn more button is clicked on the stats reverted banner.
  - `DASHBOARD_NEW_STATS_AVAILABILITY_BANNER_CANCEL_TAPPED` - when the user opts out of the new stats UI by clicking on the `No thanks` button.
  - `DASHBOARD_NEW_STATS_AVAILABILITY_BANNER_TRY_TAPPED` - when the user opts in to the new stats UI by clicking on the `Try it out` button.

**NOTE: The events have been added to the Events spreadsheet and need to be validated before the PR is approved.**

### Behaviour
#### `wc-admin` is disabled but v4 stats UI is already displayed to the user
<img width="300" src="https://user-images.githubusercontent.com/22608780/63574337-24409a00-c5a5-11e9-93eb-84bdf8342ddc.gif">

#### `wc-admin` is active but old stats UI is already displayed to the user
<img width="300" src="https://user-images.githubusercontent.com/22608780/63574199-e2afef00-c5a4-11e9-9eeb-54f9ac57c04c.gif">

#### User opts out of the new stats UI from Settings page
<img width="300" src="https://user-images.githubusercontent.com/22608780/63573897-28b88300-c5a4-11e9-9037-17df214d414d.gif">

#### User opts in to the new stats UI from Settings page
<img width="300" src="https://user-images.githubusercontent.com/22608780/63573808-f575f400-c5a3-11e9-99e0-4f993f664671.gif">

### Screenshots
#### Settings Page:
<img width="300" src="https://user-images.githubusercontent.com/22608780/63573960-4d145f80-c5a4-11e9-9277-b28a5d249bd3.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/63573992-61f0f300-c5a4-11e9-83fb-bde1d145d247.png">

#### New Stats UI availability banner:
<img width="300" src="https://user-images.githubusercontent.com/22608780/63574039-792fe080-c5a4-11e9-8324-63c24199522e.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/63574500-86999a80-c5a5-11e9-90b9-2d4f6585cc85.png">

#### New Stats UI is reverted banner:
<img width="300" src="https://user-images.githubusercontent.com/22608780/63574608-c19bce00-c5a5-11e9-91b6-901a0abaafbb.png">

### Testing scenarios
I apologise in advance for the number of testing scenarios noted here 🙏 

#### When app first installed/user logging in to a site
- [x] If `wc-admin` active in the site:
    - [x] Display the v4 stats UI availability banner. 
        - [x] If the user opts to try the new stats, display the v4 stats UI.
        - [x] If the user does not want to try the new stats, hide the banner.
        - [x] In both cases, the banner should not be displayed again even after pull to refresh or if app is closed and opened again.
- [x] If `wc-admin` is not active on a site:
    - [x] Display the current stats (no banner should be displayed)

 #### User switches store from an site which has `wc-admin` to another site which does not
- [x] Display the old stats UI (Stats reverted banner is NOT displayed)

#### User switches store from an site which does NOT have the plugin to site which HAS the plugin
- [x] If the user had NOT previously accepted to try the new stats UI for this site:
    - [x] Display the v4 stats UI availability banner. 
        - [x] If the user opts to try the new stats, display the v4 stats UI.
        - [x] If the user does not want to try the new stats, hide the banner.
        - [x] In both cases, the banner should not be displayed again even after pull to refresh or if app is closed and opened again.
- [x] If the user had previously accepted to try the new stats UI for this site:
    - [x] The new stats UI is displayed. The banner is no longer displayed.

#### User switches store from an site which has the plugin to site which also HAS the plugin
- [x] If the user had NOT previously accepted to try the new stats UI for this site:
    - [x] Display the v4 stats UI availability banner. 
        - [x] If the user opts to try the new stats, display the v4 stats UI.
        - [x] If the user does not want to try the new stats, hide the banner.
        - [x] In both cases, the banner should not be displayed again even after pull to refresh or if app is closed and opened again.
- [x] If the user had previously accepted to try the new stats UI for this site:
    - [x] The new stats UI is displayed. The banner is no longer displayed.

#### V4 stats UI is displayed but `wc-admin` is disabled now. When users pull tor refresh
- [x] Display the stats UI reverted banner.
    - [x] If the user clicks on LEARN MORE, open the WooCommerce admin plugin page.
    - [x] If the user dismisses the dialog, hide the banner. The banner should not be displayed again even after pull to refresh or if app is closed and opened again.

#### Old stats UI is displayed but `wc-admin` is active now. When users pull tor refresh
- [x] if user has already opted in for the new stats - The new stats UI is displayed without a banner.
- [x] if user has already opted out for the new stats - The old stats UI is displayed without a banner.
 - [x] if user has NOT opted in/out for the new stats - Display the v4 stats UI availability banner. 
    - [x] If the user opts to try the new stats, display the v4 stats UI.
    - [x] If the user does not want to try the new stats, hide the banner.
    - [x] In both cases, the banner should not be displayed again even after pull to refresh or if app is closed and opened again.

#### `wc-admin` is not active on a site:
- [x] The Beta features option is NOT displayed in the Settings screen.

#### User has opted in to the new stats UI from the Settings screen
- [x] Display the new stats UI (no banner should not be displayed)

#### User has opted OUT of the new stats UI from the Settings screen
- [x] Display the stats UI reverted banner.
    - [x] If the user clicks on LEARN MORE, open the WooCommerce admin plugin page.
    - [x] If the user dismisses the dialog, hide the banner. The banner should not be displayed again even after pull to refresh or if app is closed and opened again.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

anitaa1990 committed:
Added UI for reverted stats notice i.e. when wc-admin is no longer available but v4 stats is already displayed

anitaa1990 committed:
Merge branch 'feature/stats-improvements-ui-changes' into issue/1336-stats-corner-cases

# Conflicts:
#	WooCommerce/src/main/res/layout/fragment_dashboard.xml

anitaa1990 committed:
Added UI to display new stats availability card to the top of the old stats UI

anitaa1990 committed:
Add logic to display stats reverted banner if the wc-admin is disabled for a site but the v4 stats UI is already displayed

anitaa1990 committed:
Add logic to display stats availability banner if the wc-admin is enabled for a site but the old stats UI is already displayed

anitaa1990 committed:
Added beta features option in Settings page if wc-admin is active/installed on a site

anitaa1990 committed:
Added beta features fragment UI

anitaa1990 committed:
Added logic to refresh the Stats fragment based on user preferences

anitaa1990 committed:
Update toolbar title when beta features fragment is displayed

anitaa1990 committed:
Added tracking events to the new stats banner clicks

anitaa1990 committed:
Added release notes

nbradbury commented:
I haven't reviewed the code yet, but now I see this in action I'm not sure about the experience. 

Instead of asking the user whether they want to try the new stats - which they may dismiss without reading - why not automatically switch to the new stats and tell them something like, "Welcome to the new stats!" with a Dismiss button.

We know the new stats are a big improvement so let's give them to the user without making it a choice.

@kyleaparker Thoughts?

nbradbury commented:
I'm also wondering if we need to handle the edge case of users switching from a store with `wc-admin` to one without it? In this situation the stats revert to the old way without explanation. Definitely not a big deal, but wanted to mention it.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreStatsAvailabilityCard.kt:
```diff
@@ -0,0 +1,37 @@
+
+    fun initView(listener: MyStoreStatsAvailabilityListener) {
+        my_store_availability_viewMore.setOnCheckedChangeListener { _, isChecked ->
+            if (isChecked) {
```
This would be nicer with some simple animation:

```
if (isChecked) {
    WooAnimUtils.fadeIn(my_store_availability_morePanel)
} else {
    WooAnimUtils.fadeOut(my_store_availability_morePanel)
}
```

anitaa1990 replied:
Nice one! Fixed in 463d02f

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/MainSettingsFragment.kt:
```diff
@@ -134,6 +142,11 @@ class MainSettingsFragment : androidx.fragment.app.Fragment(), MainSettingsContr
         }
 
+        textBetaFeatures.setOnClickListener {
+            // TODO: add analytic event here
```
Is this TODO something that needs to be addressed in this PR?

anitaa1990 replied:
My bad! I was supposed to add the event here. I completely missed that! Added in 0d9cf76

kyleaparker commented:
Automatically loading the new stats is what was probably proposed. From the
posts and discussions around this is didn’t sound like we could do that due
to loading a cached version of stats. If that behaviour can be changed then
I’d be all for it.

On Fri, Aug 23, 2019 at 12:15 PM Nick Bradbury <notifications@github.com>
wrote:

> I'm also wondering if we need to handle the edge case of users switching
> from a store with wc-admin to one without it? In this situation the stats
> revert to the old way without explanation. Definitely not a big deal, but
> wanted to mention it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/woocommerce/woocommerce-android/pull/1371?email_source=notifications&email_token=AAFOBKPM45AX3GVUWGDI7RDQGAELFA5CNFSM4IO4WDXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5AVHHA#issuecomment-524374940>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFOBKKF7HSS3EUP43E54A3QGAELFANCNFSM4IO4WDXA>
> .
>

nbradbury commented:
> From the posts and discussions around this is didn’t sound like we could do that due to loading a cached version of stats.

Oh, right - the issue was that we have to show the old stats because of the delay determining whether the new stats are available. I'm wondering if there's a better way to handle this, though? We could simply fade out the old stats and fade in the new with a message explaining it.

kyleaparker commented:
I think for now it’s a better experience for them to opt into the stats. Especially since this is experimental and not part of core they should be able to opt in and out. 

Once this is part of core we should find a way to auto load the new stats when they launch the app without them seeing the old ones. Maybe we can default to the new stats and fallback to the old version if they don’t have it.

anitaa1990 committed:
Added animation when store stats availability card is displayed/hidden

anitaa1990 committed:
Added event when the beta features section is tapped in settings page

anitaa1990 committed:
Fixed a bug where the fragment was not replaced after stats opt in settings was changed

anitaa1990 commented:
@nbradbury thanks for the quick review! I fixed the code changes requested. 

> I'm also wondering if we need to handle the edge case of users switching from a store with wc-admin to one without it? In this situation the stats revert to the old way without explanation. Definitely not a big deal, but wanted to mention it.

I apologise, I had written the test case wrong! In this case, the reverted banner would not be displayed. 

There was a bug where the fragment was not getting replaced when the user changed the stats settings. Fixed this in d876af9.

PR ready for another round 👍

nbradbury reviewed (approved):
:shipit:
